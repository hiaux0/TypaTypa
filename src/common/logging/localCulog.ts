const debugMode = true;
const isBrowser = true;

/**
 * TODO: WARNING and ERROR mean sth else in other loggers, but here
 * it's just a "higher level"
 */
type LogLevel = "INFO" | "DEBUG" | "VERBOSE" | "WARNING" | "ERROR";

const terminalColorMap = {
  Reset: "\x1b[0m",
  Bright: "\x1b[1m",
  Dim: "\x1b[2m",
  Underscore: "\x1b[4m",
  Blink: "\x1b[5m",
  Reverse: "\x1b[7m",
  Hidden: "\x1b[8m",
  FgBlack: "\x1b[30m",
  FgRed: "\x1b[31m",
  FgGreen: "\x1b[32m",
  FgYellow: "\x1b[33m",
  FgBlue: "\x1b[34m",
  FgMagenta: "\x1b[35m",
  FgCyan: "\x1b[36m",
  FgWhite: "\x1b[37m",
  FgGray: "\x1b[90m",
  BgBlack: "\x1b[40m",
  BgRed: "\x1b[41m",
  BgGreen: "\x1b[42m",
  BgYellow: "\x1b[43m",
  BgBlue: "\x1b[44m",
  BgMagenta: "\x1b[45m",
  BgCyan: "\x1b[46m",
  BgWhite: "\x1b[47m",
  BgGray: "\x1b[100m",
} as const;

export interface LogOptions {
  /// //////////// Log
  disableLogger?: boolean;
  /**
   * Todo: Remove this for explicit methods
   * Reason: quokka output
   */
  log?: boolean;
  /**
   * When printing arrays, whether to use `console.table` or not
   */
  logTable?: boolean;
  logLevel?: LogLevel;
  onlyVerbose?: boolean;
  /**
   * TODO
   * Don't use console.log(obj), use console.log(JSON.parse(JSON.stringify(obj))).
   */
  deepLogObjects?: boolean;
  /**
   * Unless specified by `expandGroupBasedOnString`
   * TODO: `expandGroupBasedOnString` could be renamed to sth more suitable
   */
  dontLogUnlessSpecified?: boolean;
  /**
   * Even in debug mode, only log, when explicitely set via `log`.
   */
  focusedLogging?: boolean;
  /// //////////// Error
  throwOnError?: boolean;
  /** Mark a logger as error */
  isError?: boolean;
  /// //////////// Custom output
  /** */
  scope?: string;
  prefix?: number;
  useTable?: boolean;
  terminalColor?: keyof typeof terminalColorMap;
  /// //////////// Grouping
  startGroupId?: string;
  endGroupId?: string;
  /** Only in the sense of "just one" */
  isOnlyGroup?: boolean;
  clearPreviousGroupsWhen_isOnlyGroup_True?: boolean;
  /** Specified by `expandGroupBasedOnString` */
  allGroupsCollapsedButSpecified?: boolean;
  expandGroupBasedOnString?: string;
}

let defautLogOptions: LogOptions = {
  logLevel: "VERBOSE",
  clearPreviousGroupsWhen_isOnlyGroup_True: true,
  // dontLogUnlessSpecified: true,
  focusedLogging: false,
  useTable: true,
  terminalColor: "Reset",
  // allGroupsCollapsedButSpecified: true,
  // expandGroupBasedOnString: "h",
  throwOnError: true,
};

interface BugLogOptions {
  isStart?: boolean;
  isEnd?: boolean;
  //
  color?: string;
  index?: number;
  debugger?: boolean;
}

const LogLevelMap = {
  INFO: 0,
  DEBUG: 1,
  VERBOSE: 2,
  WARNING: 3,
  ERROR: 4,
};

function isAllowedLogLevel(logOptions: LogOptions): boolean {
  if (!logOptions.logLevel) return false;
  if (!defautLogOptions.logLevel) return false;

  const isAllowed =
    LogLevelMap[logOptions.logLevel] <= LogLevelMap[defautLogOptions.logLevel];

  return isAllowed;
}

const loggerDevelopmentDebugLog: string[][] = [];
const groupId: string[] = [];
let onlyGroup: string[] = [];
let bugGroupId: string[] = [];

export class CuLogger {
  private readonly logTrail: unknown[] = [];
  public getLogTrail() {
    return this.logTrail;
  }

  constructor(private globalLogOptions: LogOptions) {}

  public setLogOptions(logOptions: LogOptions) {
    this.globalLogOptions = {
      ...this.globalLogOptions,
      ...logOptions,
    };
  }

  public overwriteDefaultLogOtpions(logOptions: LogOptions) {
    defautLogOptions = {
      ...defautLogOptions,
      ...logOptions,
    };
  }

  public setScope(scopeName: string): void {
    this.globalLogOptions.scope = scopeName;
  }

  public enableBrowserDevelopmentLogging(): void {
    if (window) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      (window as unknown).loggerDevelopmentDebugLog = loggerDevelopmentDebugLog;
    }
  }

  /**
   * 1. Setup
   * 2. ">>> Actual Log"
   */
  public debug(
    messages: [string, ...unknown[]],
    logOptions?: LogOptions,
    callback?: (...message: unknown[]) => void,
  ): unknown[] {
    if (!debugMode) return [];

    const logOpt = {
      ...defautLogOptions,
      ...this.globalLogOptions,
      ...logOptions,
    };

    if (logOpt.disableLogger === true) return [];
    if (!isAllowedLogLevel(logOpt)) return [];

    //
    /** === false, because it is explicitly set */
    if (logOpt.log === false) {
      return [];
    }

    const onlyFocusedLogging = logOpt.focusedLogging && !logOpt?.log;
    if (onlyFocusedLogging) {
      return [];
    }

    //
    const [withSubstitutions] = messages;
    let [, ...placeholderValues] = messages;
    let updatedSubstitutions = `[${logOpt.scope ?? ""}] ${withSubstitutions}`;

    if (logOpt.prefix) {
      // updatedSubstitutions = `- (${logOpt.prefix}.) - ${updatedSubstitutions}`;
      updatedSubstitutions = `${logOpt.prefix} - ${updatedSubstitutions}`;
    }

    //
    if (logOpt.deepLogObjects) {
      placeholderValues = placeholderValues.map((value) =>
        JSON.stringify(value),
      );
    }

    const messageWithLogScope = [
      updatedSubstitutions,
      ...placeholderValues,
    ] as string[];

    //
    if (logOpt.throwOnError && logOpt.isError) {
      /**
       * We console.error AND throw, because we want to keep the formatting of the console.**
       */
      console.error(...messageWithLogScope);
      throw new Error("!!! [[ERROR]] Check above message !!!");
    }

    if (logOpt.logLevel !== "VERBOSE" && logOpt.onlyVerbose) {
      return [];
    }

    //
    const isExpandGroupBasedOnString = placeholderValues.includes(
      logOpt.expandGroupBasedOnString,
    );

    if (logOpt.dontLogUnlessSpecified) {
      //
      // if (!logOpt.expandGroupBasedOnString) {
      //   console.warn("Pleace specifiy `expandGroupBasedOnString`");
      // }

      //
      const onlyLogBasedOnString = isExpandGroupBasedOnString;

      if (!onlyLogBasedOnString) {
        return [];
      }
    }

    //
    if (logOpt.startGroupId) {
      if (logOpt.allGroupsCollapsedButSpecified) {
        if (!logOpt.expandGroupBasedOnString) {
          console.warn("Pleace specifiy `expandGroupBasedOnString`");
        }

        if (isExpandGroupBasedOnString) {
          console.group();
        } else {
          console.log(...messageWithLogScope);
          console.groupCollapsed();
        }
        //
      } else {
        console.log(...messageWithLogScope);
        console.group();
      }
      groupId.push(logOpt.startGroupId);
    }

    //
    if (logOpt.isOnlyGroup) {
      if (logOpt.clearPreviousGroupsWhen_isOnlyGroup_True) {
        console.clear();
      }

      if (onlyGroup.length === 0) {
        onlyGroup.push(messageWithLogScope[0]);
        console.group(...messageWithLogScope);
        loggerDevelopmentDebugLog.push(["group", ...messageWithLogScope]);
      } else {
        onlyGroup = [];
        console.groupEnd();
        loggerDevelopmentDebugLog.push(["groupEnd", ...messageWithLogScope]);
        console.group(...messageWithLogScope);
        onlyGroup.push(messageWithLogScope[0]);
        loggerDevelopmentDebugLog.push(["group", ...messageWithLogScope]);
      }
    }
    // >>> Actual log
    if (callback) {
      callback(...messageWithLogScope);
    } else {
      if (isBrowser) {
        console.log(...messageWithLogScope);
      } else {
        // nodejs
        console.log(
          terminalColorMap[logOpt.terminalColor],
          ...messageWithLogScope,
          terminalColorMap.Reset,
        );
      }
    }
    this.logTrail.push(messageWithLogScope);
    loggerDevelopmentDebugLog.push(["log", ...messageWithLogScope]);

    if (logOpt.endGroupId !== undefined && logOpt.endGroupId === groupId[0]) {
      console.groupEnd();
    }

    //
    if (logOpt.useTable) {
      if (Array.isArray(messageWithLogScope[1]) && logOpt.logTable) {
        console.table(messageWithLogScope[1]);
      }
    }

    return messageWithLogScope;
  }

  public bug(message: string, logOptions: BugLogOptions = {}): string {
    if (!debugMode) return "";

    //
    if (logOptions.isStart) {
      console.group(message);

      bugGroupId.push(message);
    }

    let finalMessage: string;
    if (logOptions.index) {
      finalMessage = `[(${logOptions.index})]: ${message}`;
    } else {
      finalMessage = message;
    }

    //
    console.log(
      `%c ${finalMessage}`,
      `background: ${logOptions.color ?? "blue"}`,
    );

    //
    const isEnd =
      logOptions.isEnd && message === bugGroupId[bugGroupId.length - 1];
    if (isEnd) {
      console.groupEnd();
      bugGroupId = bugGroupId.slice(0, bugGroupId.length - 1);
    }

    //
    if (logOptions.debugger) {
      // eslint-disable-next-line no-debugger
      debugger;
    }

    return finalMessage;
  }

  public todo(
    message: string,
    callback?: (...message: unknown[]) => void,
  ): string {
    const todoMessage = `>>>> [TODO]: %c${message}`;
    const finalMsg = [todoMessage, `background: ${"darkgreen"}`];

    if (callback) {
      callback(...finalMsg);
    } else {
      console.log(...finalMsg);
    }

    return todoMessage;
  }
}

export const logger = new CuLogger(defautLogOptions);

// export const testLogger = new Logger({
//   scope: 'Test',
//   // log: false,
// });
// testLogger.debug(['first']);
// testLogger.debug(['another'], { log: true });

/**
 * - [x?] Throw on first error
 *   - is this part of a logger?
 *   - some other debugging mode then?
 *
 * - colorize logs
 *
 * - [ ] only log based on given error codes
 * - [ ] discard "old logs"
 *   - requires separate log files for when a log was added
 */
