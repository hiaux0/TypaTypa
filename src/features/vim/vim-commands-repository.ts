import { ModifiersType } from "../../common/modules/keybindings/app-keys";

export enum VIM_COMMAND {
  "backspace" = "backspace",
  "cancelAll" = "cancelAll",
  "clearLine" = "clearLine",
  "click" = "click",
  "createNewLine" = "createNewLine",
  "createNewLineAbove" = "createNewLineAbove",
  "copy" = "copy",
  "cursorRight" = "cursorRight",
  "cursorUp" = "cursorUp",
  "cursorLeft" = "cursorLeft",
  "cursorDown" = "cursorDown",
  "cursorWordForwardEnd" = "cursorWordForwardEnd",
  "cursorWordForwardStart" = "cursorWordForwardStart",
  "cursorBackwordsStartWord" = "cursorBackwordsStartWord",
  "cursorLineEnd" = "cursorLineEnd",
  "cursorLineStart" = "cursorLineStart",
  "delete" = "delete",
  "deleteInnerWord" = "deleteInnerWord",
  "deleteLine" = "deleteLine",
  "enterInsertAfterModeEnd" = "enterInsertAfterModeEnd",
  "enterInsertAfterMode" = "enterInsertAfterMode",
  "enterInsertMode" = "enterInsertMode",
  "enterInsertModeStart" = "enterInsertModeStart",
  "enterNormalMode" = "enterNormalMode",
  "enterVisualMode" = "enterVisualMode",
  "toggleFold" = "toggleFold",
  "goToFirstLine" = "goToFirstLine",
  "goToLastLine" = "goToLastLine",
  "indentRight" = "indentRight",
  "indentLeft" = "indentLeft",
  "joinLine" = "joinLine",
  "jumpNextBlock" = "jumpNextBlock",
  "jumpPreviousBlock" = "jumpPreviousBlock",
  "newLine" = "newLine",
  "paste" = "paste",
  "pasteVim" = "pasteVim",
  "redo" = "redo",
  "repeatLastCommand" = "repeatLastCommand",
  "replace" = "replace",
  "snippet" = "snippet",
  "space" = "space",
  "toCharacterAtBack" = "toCharacterAtBack",
  "toCharacterAt" = "toCharacterAt",
  "toCharacterAfterBack" = "toCharacterAfterBack",
  "toCharacterBefore" = "toCharacterBefore",
  "type" = "type",
  "undo" = "undo",
  "unfoldAll" = "unfoldAll",
  "yank" = "yank",
  "visualDelete" = "visualDelete",
  "visualInnerWord" = "visualInnerWord",
  "visualInnerBlock" = "visualInnerBlock",
  "visualStartLineWise" = "visualStartLineWise",
  "visualMoveToOtherEndOfMarkedArea" = "visualMoveToOtherEndOfMarkedArea",
  // custom
  "toggleCheckbox" = "toggleCheckbox",
  "hint" = "hint",
  "shift" = "shift",
  "customExecute" = "customExecute",
  "nothing" = "nothing",
}

export const VIM_MODE_COMMANDS = [
  VIM_COMMAND["enterInsertMode"],
  VIM_COMMAND["enterInsertModeStart"],
  VIM_COMMAND["enterNormalMode"],
  VIM_COMMAND["enterVisualMode"],
  VIM_COMMAND["visualStartLineWise"],
  VIM_COMMAND["createNewLine"],
  VIM_COMMAND["createNewLineAbove"],
  VIM_COMMAND["enterInsertAfterModeEnd"],
  VIM_COMMAND["enterInsertAfterMode"],
];

export const VIM_COMMANDS = [
  VIM_COMMAND.newLine,
  VIM_COMMAND["backspace"],
  VIM_COMMAND["createNewLine"],
  VIM_COMMAND["createNewLineAbove"],
  VIM_COMMAND["cursorRight"],
  VIM_COMMAND["cursorUp"],
  VIM_COMMAND["cursorLeft"],
  VIM_COMMAND["cursorDown"],
  VIM_COMMAND["cursorWordForwardEnd"],
  VIM_COMMAND["cursorWordForwardStart"],
  VIM_COMMAND["cursorBackwordsStartWord"],
  VIM_COMMAND["cursorLineEnd"],
  VIM_COMMAND["cursorLineStart"],
  VIM_COMMAND["delete"],
  VIM_COMMAND["deleteLine"],
  VIM_COMMAND["deleteInnerWord"],

  VIM_COMMAND["indentRight"],
  VIM_COMMAND["indentLeft"],
  VIM_COMMAND["joinLine"],
  VIM_COMMAND["jumpNextBlock"],
  VIM_COMMAND["jumpPreviousBlock"],
  VIM_COMMAND["newLine"],
  VIM_COMMAND["paste"],
  VIM_COMMAND["replace"],
  VIM_COMMAND["space"],
  VIM_COMMAND["toCharacterAtBack"],
  VIM_COMMAND["toCharacterAt"],
  VIM_COMMAND["toCharacterAfterBack"],
  VIM_COMMAND["toCharacterBefore"],
  VIM_COMMAND["type"],
  VIM_COMMAND["unfoldAll"],
  VIM_COMMAND["yank"],
  // visual
  VIM_COMMAND["visualDelete"],
  VIM_COMMAND["visualInnerWord"],
  VIM_COMMAND["visualInnerBlock"],
  VIM_COMMAND["visualStartLineWise"],
  VIM_COMMAND["visualMoveToOtherEndOfMarkedArea"],
  // ...VIM_MODE_COMMANDS,
  VIM_COMMAND["enterInsertMode"],
  VIM_COMMAND["enterInsertModeStart"],
  VIM_COMMAND["enterNormalMode"],
  VIM_COMMAND["enterVisualMode"],
];
export type VimCommandNames = keyof typeof VIM_COMMAND;

export const VIM_COMMANDS_THAT_CHANGE_TO_NORMAL_MODE = ["visualDelete"];

export interface VimCommand {
  /**
   * Modifiers: <mod>
   * Full: <modA><modB>keyAkeyB
   */
  key: string;
  command?: VimCommandNames;
  sequence?: string;
  execute?: () => void;
  args?: {
    text?: string;
  };
  desc?: string;
}

export interface SynonymKey {
  [key: string]: ModifiersType;
}
