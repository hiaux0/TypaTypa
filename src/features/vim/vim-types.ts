import { ISnippet } from "../../common/modules/keybindings/snippets";
import { getRandomId } from "../../common/modules/random/random";
import { EditorId } from "../../domain/types/types";
import {
  VimCommand,
  SynonymKey,
  VIM_COMMAND,
  VimCommandNames,
} from "./vim-commands-repository";
import { VimCore } from "./vimCore/VimCore";

type VimCore_TODO = "todo vim Core";

export interface KeyBindingModes {
  insert?: VimCommand[];
  [VimMode.INSERT]?: VimCommand[];
  normal?: VimCommand[];
  [VimMode.NORMAL]?: VimCommand[];
  visual?: VimCommand[];
  [VimMode.VISUAL]?: VimCommand[];
  visualline?: VimCommand[];
  [VimMode.VISUALLINE]?: VimCommand[];
  synonyms?: SynonymKey;
}

export enum VimExecutingMode {
  "INDIVIDUAL" = "INDIVIDUAL",
  "BATCH" = "BATCH",
}

/**
 * Given a string 'Hello, World'
 * And I'm in normal mode
 * When I type "l"
 * Then the cursor should move one right
 */
export interface FindPotentialCommandReturn {
  targetCommand: VimCommand | undefined;
  potentialCommands: VimCommand[];
  keySequence: string;
}

export type IndentationLevel = number;
export type Text = string;

export interface IndentationNode {
  text?: Text;
  indentation?: IndentationLevel;
}

export type LineId = string;
export interface VimLine extends IndentationNode {
  text: Text;
  id?: LineId; // Migration_1
  // cursor?: Cursor;
}

export const EMPTY_VIM_LINE: VimLine = { text: "", id: getRandomId() };

export type FoldMap = Record<string, boolean>;

export interface IVimState {
  id: EditorId;
  cursor?: Cursor;
  lines?: VimLine[];
  mode?: VimMode;
  foldMap?: FoldMap;
  visualStartCursor?: Cursor;
  visualEndCursor?: Cursor;
  deletedLinesIndeces?: number[];
  /**
   * {
   *   "\"": [...],
   *   "+": [...],
   *   "*": [...],
   * }
   */
  deletedLines?: Record<string, VimLine[]>;
  /** * Should live in key mapping/binding? */
  commandName?: VimCommandNames;
  /** * Should live in key mapping/binding? */
  snippet?: ISnippet;
}

export interface QueueInputReturn {
  vimState: IVimState | null | undefined;
  targetCommand: VIM_COMMAND;
  keys: string;
}

/**
 * 0 based index
 */
export interface Cursor {
  col: number;
  line: number;
}
export enum VimMode {
  "NORMAL" = "NORMAL",
  "INSERT" = "INSERT",
  "VISUAL" = "VISUAL",
  "VISUALLINE" = "VISUALLINE",
}
export type VimModeKeys = keyof typeof VimMode;

export interface VimHooks {
  afterInit?: (
    vim: VimCore,
  ) => QueueInputReturn[] | Promise<QueueInputReturn[]> | void;
  onBeforeCommand?: () => boolean;
  commandListener?: CommandListener;
  modeChanged?: ModeChanged;
  vimStateUpdated?: (vimState: IVimState) => void;
}

export interface VimOptions {
  vimState?: IVimState;
  keyBindings?: KeyBindingModes;
  container?: HTMLElement;
  childSelector?: string;
  leader?: string;
  vimPlugins?: VimPlugin[];
  indentSize?: number;
  hooks?: VimHooks;
}

export interface VimPlugin {
  commandName: string;
  execute: (vimState?: IVimState, commandValue?: unknown) => IVimState | void;
}

export interface InputData {
  pressedKey: string;
  ev: KeyboardEvent;
  modifiersText: string;
}

export type CommandListener = (
  vimResult: QueueInputReturn,
  // inputData: InputData,
  // vim?: VimCore
) => void | IVimState;
export type ModeChanged = (
  vimResults: QueueInputReturn,
  newMode?: VimMode,
  oldMode?: VimMode,
  vim?: VimCore_TODO,
) => void | IVimState;

export interface VimEditorOptions {
  vimState?: IVimState;
  id?: EditorId;
  startCursor?: Cursor;
  startLines?: VimLine[];
  container?: HTMLElement;
  caret?: HTMLElement;
  childSelector?: string;
  removeTrailingWhitespace?: boolean;
  plugins?: VimPlugin[];
  afterInit?: (
    vim: VimCore_TODO,
  ) => QueueInputReturn[] | Promise<QueueInputReturn[]> | void;
  onBeforeCommand?: () => boolean;
  commandListener?: CommandListener;
  modeChanged?: ModeChanged;
  vimStateUpdated?: (vimState: IVimState) => void;
}
