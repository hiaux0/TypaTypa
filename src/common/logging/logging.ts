import { CuLogger as Culogger, LogOptions } from "./localCulog";

interface ILogOptions extends LogOptions {
  log?: boolean;
  focusedLogging?: boolean;

  measurePerf?: boolean;
  focusedPerf?: boolean;
  logPerf?: boolean;

  /**
   * Indicate to, that log can/should perform a reset
   */
  reset?: boolean;
  highlight?: boolean;
}

export function logOptionsGuard(input: any): input is ILogOptions {
  if (typeof input !== "object") {
    return false;
  }

  const keys = Object.keys(input);
  const validKeys = [
    "log",
    "focusedLogging",
    "measurePerf",
    "focusedPerf",
    "logPerf",
    "reset",
    "highlight",
  ];
  for (const key of keys) {
    if (!validKeys.includes(key)) {
      return false;
    }
  }
  return true;
}

const DEFAULT_LOG_OPTIONS: ILogOptions = {
  log: false,
  focusedLogging: true,
  measurePerf: false,
  focusedPerf: true,
  logPerf: false,
  reset: false,
};

export class Logger {
  public readonly culogger: Culogger;

  constructor(
    scope = "Aurelia",
    private readonly options: ILogOptions = DEFAULT_LOG_OPTIONS,
  ) {
    this.culogger = new Culogger({ scope });
    this.culogger.overwriteDefaultLogOtpions({
      logTable: true,
      logLevel: "INFO",
      focusedLogging: false,
      ...options,
      // logScope: false,
    });
  }

  log(...args: any[]) {
    let options = {} as ILogOptions;
    if (args.length > 2 && typeof args[args.length - 1] === "object") {
      options = args.pop() as ILogOptions;
    }
    const message = args.join(" ");

    /**
     * Wallaby logic.
     * Wallaby does not console.log from external library.
     */
    this.logMessage(message, options);
  }

  private logMessage(
    message: string,
    options: ILogOptions = DEFAULT_LOG_OPTIONS,
  ) {
    const { log } = options;
    const loggedMessage = this.culogger.debug([message], {
      logLevel: "INFO",
      log,
      ...options,
    });
    if (loggedMessage.length > 0) {
      // console.log(loggedMessage[0]);
      if (loggedMessage.length > 1) {
        console.log("There are more log messages");
      }
    }
  }
}
