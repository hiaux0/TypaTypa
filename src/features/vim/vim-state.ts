import { VimCommandNames } from "./vim-commands-repository";
import {
  Cursor,
  EMPTY_VIM_LINE,
  FoldMap,
  VimLine,
  VimMode,
  IVimState,
} from "./vim-types";
import { ISnippet } from "../../common/modules/keybindings/snippets";
import { EditorId } from "../../domain/types/types";
import generateId from "../../common/modules/random/random";
import { Logger } from "../../common/logging/logging";
 
const logger = new Logger("VimState");

export class VimStateClass {
  id: EditorId;
  cursor: Cursor | undefined;
  foldMap: FoldMap | undefined;
  lines: VimLine[] | undefined;
  mode: VimMode | undefined;
  visualStartCursor: Cursor | undefined;
  visualEndCursor: Cursor | undefined;
  deletedLinesIndeces: number[] | undefined;
  deletedLines?: IVimState["deletedLines"] = {};
  commandName: VimCommandNames | undefined;
  snippet: ISnippet | undefined;

  constructor(private readonly vimState: IVimState) {
    if (!vimState.id) {
      vimState.id = generateId();
    }
    this.updateVimState(vimState);
  }

  public static create(cursor: Cursor, lines?: VimLine[]) {
    return new VimStateClass({ cursor, lines, id: "" });
  }

  public static createEmpty() {
    return new VimStateClass({
      cursor: { col: 0, line: 0 },
      lines: [{ text: "" }],
      id: "",
    });
  }

  public serialize(): IVimState {
    return {
      id: this.id,
      cursor: this.cursor,
      foldMap: this.foldMap,
      lines: this.lines,
      mode: this.mode,
      visualStartCursor: this.visualStartCursor,
      visualEndCursor: this.visualEndCursor,
      deletedLinesIndeces: this.deletedLinesIndeces,
      deletedLines: this.deletedLines,
      commandName: this.commandName,
      snippet: this.snippet,
    };
  }

  public getLineAt(lineIndex: number) {
    if (!this.lines) return;

    const line = this.lines[lineIndex];
    return line;
  }

  public getActiveLine() {
    if (!this.cursor) return;
    const active = this.getLineAt(this.cursor.line) ?? EMPTY_VIM_LINE;
    // const active = this.getLineAt(this.cursor.line);
    return active;
  }

  public getPreviousLine() {
    if (!this.cursor) return;
    const lineIndex = Math.max(this.cursor.line - 1, 0);
    const previous = this.getLineAt(lineIndex);
    return previous;
  }

  public getNextLine() {
    if (!this.cursor) return;
    if (!this.lines) return;

    const nextIndex = this.getNextCursorLine();
    if (!nextIndex) return;
    const previous = this.getLineAt(nextIndex);

    return previous;
  }

  public getNextCursorLine() {
    if (!this.cursor) return;
    if (!this.lines) return;

    const lineLength = this.lines.length;
    const next = Math.min(this.cursor.line + 1, lineLength - 1);

    return next;
  }

  public insertLine(lineIndex: number, newLine: VimLine) {
    const tempLines = structuredClone(this.lines);
    if (!tempLines) return;
    tempLines.splice(lineIndex, 0, newLine);
    return tempLines;
  }

  public updateLine(lineIndex: number, updated: string) {
    if (!this.lines) return;
    this.lines[lineIndex] = { text: updated };
  }

  public updateActiveLine(updated: string) {
    if (!this.cursor) return;
    this.updateLine(this.cursor.line, updated);
  }

  public updateCursor(cursor: Cursor) {
    this.cursor = cursor;
  }

  public updateVimState(vimState: IVimState) {
    Object.assign(this, vimState);
    this.mode = vimState.mode ?? VimMode.NORMAL;
  }

  public reportVimState() {
    const { cursor, lines, mode } = this;
    if (!lines) return;
    logger.culogger.overwriteDefaultLogOtpions({ log: true });
    /* prettier-ignore */ if (mode) logger.culogger.debug(['Vim in Mode:', mode], {}, (...r) => console.log(...r));
    /* prettier-ignore */ logger.culogger.debug(['Cursor at', {...cursor}], {}, (...r) => console.log(...r));
    /* prettier-ignore */ logger.culogger.debug(['Lines are', lines.map(l => l.text)], {}, (...r) => console.log(...r));
    logger.culogger.overwriteDefaultLogOtpions({ log: false });
  }

  public isInsertMode(): boolean {
    const is = this.mode === VimMode.INSERT;
    return is;
  }
}
