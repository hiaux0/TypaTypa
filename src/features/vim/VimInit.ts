import { Logger } from "../../common/logging/logging";
import { VIM_COMMAND } from "./vim-commands-repository";
import { IVimState, VimOptions } from "./vim-types";
import { VimInputHandler } from "./VimInputHandler";

const logger = new Logger("VimInit");

export class VimInit {
  private vimInputHandler: VimInputHandler;
  private options: VimOptions;

  public init(options?: VimOptions) {
    this.vimInputHandler = new VimInputHandler();
    this.vimInputHandler.init(options);

    if (!options) return;
    this.options = options;
    this.vimInputHandler;
  }

  public executeCommandSequence(sequence: string): void {
    if (!this.vimInputHandler) console.log("Please call #init() first");
    this.vimInputHandler.executeCommandSequence(sequence);
  }

  public executeCommand(
    commandName: VIM_COMMAND,
    inputForCommand: string,
  ): IVimState | undefined {
    if (!this.vimInputHandler)
      throw "[ERROR:VimInit] Please call #init() first";
    const result = this.vimInputHandler.executeCommand(
      commandName,
      inputForCommand,
    );
    return result;
  }

  public reload(vimState: IVimState) {
    this.vimInputHandler.reload(vimState);
  }

  public getOptions() {
    return this.options;
  }
}
