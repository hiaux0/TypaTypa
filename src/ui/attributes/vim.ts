import { INode, ISignaler, resolve } from "aurelia";
import { VimStateClass } from "../../features/vim/vim-state";
import { IVimState, VimOptions } from "../../features/vim/vim-types";
import { VimInit } from "../../features/vim/VimInit";
import { IVimInputHandlerV2 } from "../../features/vim/VimInputHandlerV2";

export class VimCustomAttribute {
  private element = resolve(INode) as HTMLElement;
  vimState: IVimState;

  constructor(
    private vimInit: VimInit = resolve(VimInit),
    private vimInputHandler: IVimInputHandlerV2 = resolve(IVimInputHandlerV2),
  ) {}

  attached() {
    console.log(this.element);

    const vimState = VimStateClass.createEmpty();
    vimState.id = "bruh";
    vimState.lines = [{ text: "hi, world what are" }];
    const options: VimOptions = {
      vimState,
      vimId: vimState.id,
      container: this.element,
      childSelector: "vim-line",
      hooks: {
        vimStateUpdated: (vimState: IVimState) => {
          this.vimState = vimState;
        },
      },
    };
    this.vimInit.init(options);
    this.vimInputHandler.registerAndInit(options);
    this.vimState = vimState;
  }
}
