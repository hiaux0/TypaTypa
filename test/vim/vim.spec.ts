import { describe, expect, test } from "vitest";
import { VimInit } from "../../src/features/vim/VimInit";
import { VIM_COMMAND } from "../../src/features/vim/vim-commands-repository";
import { Logger } from "../../src/common/logging/logging";
import { VimStateClass } from "../../src/features/vim/vim-state";
import { KeyMappingService } from "../../src/features/vim/vimCore/commands/KeyMappingService";
import { KeyBindingModes, VimMode } from "../../src/features/vim/vim-types";

const logger = new Logger("VimInit.spec.ts", { terminalColor: "FgMagenta" });

describe("VimInit", () => {
  test("Init", () => {
    const vimInit = new VimInit();
    vimInit.init();
    const result = vimInit.executeCommand(VIM_COMMAND.enterInsertMode, "");
    expect(result).toMatchSnapshot();
  });

  test.only("Cursor", () => {
    const mappingByKey = {};
    const mappingByMode: KeyBindingModes = {
      [VimMode.NORMAL]: [
        {
          key: "a",
          execute: () => {
            return true;
          },
        },
      ],
    };
    new KeyMappingService().init(mappingByKey, mappingByMode);
    const keyBindings = KeyMappingService.keyBindings;
    const vimInit = new VimInit();
    const vimState = VimStateClass.createEmpty();
    vimState.lines = [{ text: "Hello, World" }];
    vimInit.init({ vimState, keyBindings });

    const result = vimInit.executeCommand(VIM_COMMAND.customExecute, "a");
    /*prettier-ignore*/ logger.culogger.debug(["[vim.spec.ts,20] result: ", result]);
    // expect(result).toMatchSnapshot();
    // expect(false).toBe(true);
  });
});

/**
 * 
1. Press key
2. Map to command
3. Execute command
4. Update state
 */
