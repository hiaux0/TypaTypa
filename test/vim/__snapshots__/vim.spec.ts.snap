// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VimInit > Cursor 1`] = `
{
  "cursor": {
    "col": 0,
    "line": 0,
  },
  "deletedLines": {},
  "id": "8iorzxt-b88n30hk20",
  "lines": [
    {
      "text": "",
    },
  ],
  "mode": "NORMAL",
}
`;
